File: backend/src/test/java/com/onlinejudge/common/GatewayRoutingContractTest.java
package com.onlinejudge.common;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;

class GatewayRoutingContractTest {
    private static final Path TEMPLATE = Path.of("..", "deploy", "gateway", "gateway.conf.template");
    private static final Path REQUEST_HEADERS = Path.of("..", "deploy", "gateway", "proxy-request-headers.conf");

    @Test
    void gatewayRoutesEveryFrozenPublicApiFamilyWithoutCoursePrefixConflicts() throws IOException {
        String template = Files.readString(TEMPLATE);

        assertThat(template).contains(
                "location = /api/v1/courses",
                "location /api/v1/courses/",
                "location = /api/v1/homeworks",
                "location ^~ /api/v1/homeworks/",
                "location = /api/v1/notifications",
                "location ^~ /api/v1/notifications/",
                "location = /api/v1/reminder-rules",
                "location = /api/v1/grades",
                "location ^~ /api/v1/grade-items/",
                "location ^~ /api/v1/grade-records/",
                "location ^~ /api/v1/course-grade-summaries/",
                "location ^~ /api/v1/grade-review-requests/",
                "/(grades|grade-items|grade-rules|grade-publish-records|grade-change-logs|my-grades|grade-analysis|grade-review-requests|my-grade-review-requests)(/|$)");
        assertThat(template.indexOf("/(labs|homeworks)"))
                .isLessThan(template.indexOf("location /api/v1/courses/"));
        assertThat(template.indexOf("/(grades|grade-items|grade-rules"))
                .isLessThan(template.indexOf("location /api/v1/courses/"));
        assertThat(template).doesNotContain("location ^~ /api/v1/courses/");
    }

    @Test
    void gatewayForwardsBearerButClearsAllBrowserSuppliedIdentityHeaders() throws IOException {
        String requestHeaders = Files.readString(REQUEST_HEADERS);

        assertThat(requestHeaders).contains(
                "proxy_pass_request_headers off;",
                "proxy_set_header Authorization $http_authorization;",
                "proxy_set_header X-Request-Id $gateway_request_id;");
        assertThat(requestHeaders).doesNotContain(
                "X-User-Id",
                "X-Username",
                "X-User-Role",
                "X-Permissions",
                "X-Course-Ids",
                "X-Manageable-Course-Ids",
                "X-OnlineJudge-Service-Authorization",
                "X-Internal-Token");
    }

    @Test
    void gatewayDefinesUploadTimeoutFailureAndSpaBoundaries() throws IOException {
        String template = Files.readString(TEMPLATE);

        assertThat(template).contains(
                "client_max_body_size 10m;",
                "client_max_body_size 55m;",
                "proxy_connect_timeout 5s;",
                "proxy_read_timeout 60s;",
                "proxy_send_timeout 60s;",
                "proxy_read_timeout 300s;",
                "proxy_send_timeout 300s;",
                "proxy_next_upstream off;",
                "error_page 413 = @gateway_payload_too_large;",
                "error_page 429 = @gateway_rate_limited;",
                "error_page 502 = @gateway_bad_gateway;",
                "error_page 503 = @gateway_unavailable;",
                "error_page 504 = @gateway_timeout;",
                "\"code\":\"GATEWAY_413\"",
                "\"code\":\"GATEWAY_429\"",
                "\"code\":\"GATEWAY_502\"",
                "\"code\":\"GATEWAY_503\"",
                "\"code\":\"GATEWAY_504\"",
                "location ^~ /internal/v2/",
                "location ^~ /api/",
                "proxy_pass http://frontend:80;");
        assertThat(template).doesNotContain("backend:8080", "try_files $uri /index.html;");
    }
}

File: scripts/gateway/tests/gateway-routing-contract.test.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

function read(relativePath) {
  return readFileSync(fileURLToPath(new URL(relativePath, import.meta.url)), "utf8");
}

const manifest = JSON.parse(read("../../../deploy/platform/workloads.json"));
const template = read("../../../deploy/gateway/gateway.conf.template");

const expected = [
  ["identity-service", 8081, "IDENTITY"],
  ["course-service", 8082, "COURSE"],
  ["assessment-api", 8083, "ASSESSMENT"],
  ["grade-service", 8084, "GRADE"],
];

assert.equal(manifest.workloads.length, 9, "#306 freezes exactly nine workloads");
assert.equal(
  manifest.workloads.some((entry) => entry.name === "learning-service"),
  false,
  "Learning must be hosted by Course",
);

for (const [name, port, token] of expected) {
  const workload = manifest.workloads.find((entry) => entry.name === name);
  assert.ok(workload, `${name} must exist in the workload manifest`);
  assert.equal(workload.ports[0].containerPort, port, `${name} port drifted`);
  assert.match(
    template,
    new RegExp(`proxy_pass http://__${token}_UPSTREAM__`),
    `${name} needs a public proxy route`,
  );
}

const requiredRoutes = [
  "location ^~ /api/v1/auth/",
  "location = /api/v1/users/me",
  "location ^~ /api/v1/admin/",
  "location = /api/v1/courses",
  "location ^~ /api/v1/chapters/",
  "location ^~ /api/v1/labs/",
  "location = /api/v1/homeworks",
  "location ^~ /api/v1/submissions/",
  "location ^~ /api/v1/evaluations/",
  "location ^~ /api/v1/grades/",
  "location ^~ /api/v1/grade-items/",
  "location ^~ /api/v1/grade-records/",
  "location ^~ /api/v1/learning/",
  "location = /api/v1/notifications",
  "location = /api/v1/reminder-rules",
];
for (const route of requiredRoutes) {
  assert.ok(template.includes(route), `missing public route: ${route}`);
}

for (const rejectedPrefix of ["location ^~ /internal/v2/ {", "location ^~ /api/ {"]) {
  const start = template.indexOf(rejectedPrefix);
  assert.ok(start >= 0, `missing rejected prefix: ${rejectedPrefix}`);
  const end = template.indexOf("\n    }", start);
  assert.ok(template.slice(start, end).includes("return 404"), `${rejectedPrefix} must return 404`);
}
assert.doesNotMatch(template, /backend:8080|LEARNING_GRADE|__AUTH_UPSTREAM__|__CRS_UPSTREAM__/);
assert.doesNotMatch(template, /LEARNING_UPSTREAM|__LEARNING_UPSTREAM__|learning-service/);
assert.match(template, /proxy_pass_request_headers off|proxy-request-headers\.conf/);
assert.match(template, /proxy_next_upstream off;/);
assert.match(template, /error_page 413 = @gateway_payload_too_large;/);
assert.match(template, /error_page 429 = @gateway_rate_limited;/);
assert.match(template, /error_page 502 = @gateway_bad_gateway;/);
assert.match(template, /error_page 503 = @gateway_unavailable;/);
assert.match(template, /error_page 504 = @gateway_timeout;/);
assert.match(template, /add_header X-Request-Id \$gateway_request_id always;/);

const assessmentCourseRoute = template.indexOf(
  "location ~ ^/api/v1/courses/[0-9]+/(labs|homeworks)(/|$)",
);
const gradeCourseRoute = template.indexOf(
  "location ~ ^/api/v1/courses/[0-9]+/(grades|grade-items|grade-rules|grade-publish-records|grade-change-logs|my-grades|grade-analysis|grade-review-requests|my-grade-review-requests)(/|$)",
);
const genericCourseRoute = template.indexOf("location /api/v1/courses/");
assert.ok(assessmentCourseRoute >= 0, "missing Assessment course subresource route");
assert.ok(gradeCourseRoute >= 0, "missing Grade course subresource route");
assert.ok(genericCourseRoute > assessmentCourseRoute, "Course route must follow Assessment route");
assert.ok(genericCourseRoute > gradeCourseRoute, "Course route must follow Grade route");

for (const route of [
  "location ^~ /api/v1/learning/",
  "location = /api/v1/notifications",
  "location ^~ /api/v1/notifications/",
  "location = /api/v1/reminder-rules",
  "location ^~ /api/v1/reminder-rules/",
]) {
  const start = template.indexOf(route);
  assert.ok(start >= 0, `missing Course-owned route: ${route}`);
  const end = template.indexOf("\n    }", start);
  assert.match(
    template.slice(start, end),
    /proxy_pass http:\/\/__COURSE_UPSTREAM__;/,
    `${route} must route to Course`,
  );
}

console.log("gateway-routing-contract.test: PASS (services=4)");
